docs/containers.conf.5.md

```diff
@@ -195,13 +197,23 @@ Pass all host environment variables into the container.
 
 **host_containers_internal_ip**=""
 
-Set the ip for the host.containers.internal entry in the containers /etc/hosts
-file. This can be set to "none" to disable adding this entry. By default it
-will automatically choose the host ip.
+Set the IP address the container should expect to connect to the host. The IP
+address is used by Podman to automatically add the `host.containers.internal`
+and `host.docker.internal` hostnames to the container's `/etc/hosts` file. It
+is also used for the *host-gateway* flag of Podman's `--add-host` CLI option.
+If no IP address is configured (the default), Podman will try to determine it
+automatically, but might fail to do so depending on the container's network
+setup. Adding the internal hostnames to `/etc/hosts` is silently skipped then.
+Set this config to `none` to never add the internal hostnames to `/etc/hosts`.
 
-NOTE: When using podman machine this entry will never be added to the containers
-hosts file instead the gvproxy dns resolver will resolve this hostname. Therefore
-it is not possible to disable the entry in this case.
+Note: This config doesn't affect the actual network setup, it just tells Podman
+the IP address it should expect. Configuring an IP address here doesn't ensure
+that the container can actually reach the host using this IP address.
+
+Note: This config has no effect with `podman machine`, because Podman isn't
+modifying the guest's `/etc/hosts` file. The `host.containers.internal` and
+`host.docker.internal` hostnames are instead resolved by the gvproxy DNS
+resolver. Therefore it is not possible to disable the hostnames in this case.
```

Comment:
So yes this should say setting this to none has no effect, and notably with podman machine we do not add this as host entry but rather let gvproxy resolve it as dns name which is different.

Also for reference this doesn't work today: https://github.com/containers/podman/issues/21681

Reply:
I see... Just to get the expected behaviour straight (I've never used `podman machine` myself):

- [containers/podman#21681](https://github.com/containers/podman/issues/21681) is indeed a bug that will be resolved at some point, i.e. *host-gateway* not working with `podman machine` is not intended. Therefore we must not mention this in the docs of `--add-host`. Correct?

- Just to clarify, Podman's `--add-host` is always modifying the `/etc/hosts` file, even with `podman machine` (i.e. gvproxy isn't involved with `--add-host`), correct?

- `host_containers_internal_ip="none"` not working with `podman machine` is the intended behaviour and there are no plans to support it (i.e. it's no bug that will be resolved at some point)? I don't know the code (like, at all), so please excuse me if I'm asking stupid questions, but since gvproxy already reads `host_containers_internal_ip`, why can't it interpret the `"none"` value? It seems a bit odd from an user's perspective. There might be good technical reasons though (so please don't feel obligated to explain them in detail, just "it's intended" fully satisfies me and I'll update the docs accordingly).

Reply:
> https://github.com/containers/podman/issues/21681 is indeed a bug that will be resolved at some point, i.e. host-gateway not working with podman machine is not intended. Therefore we must not mention this in the docs of --add-host. Correct?

Yes that should be fixed at some point.

> Just to clarify, Podman's --add-host is always modifying the /etc/hosts file, even with podman machine (i.e. gvproxy isn't involved with --add-host), correct?

Correct

> host_containers_internal_ip="none" not working with podman machine is the intended behaviour and there are no plans to support it (i.e. it's no bug that will be resolved at some point)? I don't know the code (like, at all), so please excuse me if I'm asking stupid questions, but since gvproxy already reads host_containers_internal_ip, why can't it interpret the "none" value?

gvproxy is running on the host and proxies the network between host and VM, it doesn't read containers.conf at all so it has no idea about this entry as there is no process inside the VM
With machine we try to ensure host is the actual VM host OS (i.e. macos) so if user try to make connection they actually reach other services running on host OS.
Now because the podman in the VM has no idea what the real host ip addresses are I decided to simply skip this entry in that case and let gvproxy (which at that time already resolved this dns name) resolve it as it was the easiest way to go about.

I am not sure I would call this intended but I do not see a way to avoid that. "none" means to not write the entry and because machine already does not write it and depends on dns there is no difference in practise between the two with podman machine.

Reply:
Ah, now I get it :bulb:

I had the misconception that gvproxy is reading `host_containers_internal_ip` from `containers.conf`. I got this false impression from ["if you set a specific ip it should add it correctly"](https://github.com/containers/common/pull/2184#issuecomment-2386440253), because I assumed that "it" meant gvproxy since you were talking about gvproxy there, but you actually meant `/etc/hosts`.

So, to summarize: gvproxy isn't reading `host_containers_internal_ip` at all. However, setting `host_containers_internal_ip` to an IP address still causes Podman to write that IP address to `/etc/hosts`. This `/etc/hosts` entry then causes most software to bypass gvproxy; however, if one would query gvproxy directly (e.g. with `nslookup host.containers.internal`), gvproxy would still yield a different IP address. `host_containers_internal_ip="none"` simply has no effect because not writing anything to `/etc/hosts` can't possibly bypass gvproxy. Correct?

I've updated the docs accordingly.

I intentionally refrained from mentioning `host_containers_internal_ip="none"` in this context. The reason is that nothing special is happening there, it's just the consequence of gvproxy not reading `host_containers_internal_ip` and that only writing something to `/etc/hosts` can bypass gvproxy. I furthermore assume that very few users actually want to specifically disable the internal hostnames.